onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp

```diff
@@ -0,0 +1,259 @@
+
+        status = onert_micro::train::OMBackpropExecute::runBackward(
+          config, backprop_execute_args, backward_graph.getRuntimeAllocator());
+        assert(status == Ok);
+        if (status != Ok)
+          return status;
+      }
+
+      //  e. Update optimization state
+      {
+        status = _training_handler.updateOptimizerState(config, backward_storage, backward_context);
+        assert(status == Ok);
+        if (status != Ok)
+          return status;
+      }
+
+      //  f. Reset runtime graphs
+      backward_graph.reset();
+      forward_graph.reset();
+    }
+  }
+
+  // B - Update weights according chosen optimization strategy
+  for (uint32_t i = 0; i < _graphs.size(); ++i)
+  {
+    // Get backward context
+    OMRuntimeGraph &backward_graph = _backward_graphs.at(i);
+    OMRuntimeContext &backward_context = backward_graph.getRuntimeContext();
+    status = _training_handler.updateWeights(config, backward_context);
+  }
+
+  return status;
+}
+
+/*
+ * Evaluate metrics
+ *
+ *  Warning: 1) assume that all metric_val for all OMMetrics types actually are float values.
+ *           2) metric_val should be initialized with some value before calling this method due to
+ *              after calculation for current batch_num (the sequence number of the current sample)
+ *              this value is added to metric_val
+ */
+OMStatus OMTrainingRuntimeModule::evaluateMetric(OMMetrics metric, void *metric_val,
+                                                 uint32_t test_size)
+{
+  OMStatus status = Ok;
+  OMRuntimeGraph &forward_main_graph = _graphs.at(0);
+  uint32_t input_nums = forward_main_graph.getNumberOfInputs();
+  for (uint32_t b = 0; b < test_size; ++b)
+  {
+    //  a. Allocate forward inputs and memcopy current sample (in current batch) into forward inputs
+    {
+      forward_main_graph.allocateGraphInputs();
+      for (uint32_t i = 0; i < input_nums; ++i)
+      {
+        uint8_t *allocated_input_data =
+          reinterpret_cast<uint8_t *>(forward_main_graph.getInputDataAt(i));
+        size_t type_size = forward_main_graph.getInputDataTypeSize(i);
+        size_t tensor_size = forward_main_graph.getInputSizeAt(i);
+        size_t offset = b * type_size * tensor_size;
+
+        uint8_t *input_data = _training_handler.getInputData(i);
+        std::memcpy(allocated_input_data, input_data + offset, tensor_size * type_size);
+      }
+    }
+
+    //  b. Run forward graph
+    {
+      status = run();
+      assert(status == Ok);
+      if (status != Ok)
+        return status;
+    }
+
+    // c. Calculate loss on test_size data
+    _training_handler.evaluateMetric(metric, metric_val, forward_main_graph.getRuntimeStorage(),
+                                     forward_main_graph.getRuntimeContext(), b);
+
+    // Reset values
+    forward_main_graph.reset();
+  }
+
+  // Averaged result
+  {
+    float *f_metric_val = reinterpret_cast<float *>(metric_val);
+    *f_metric_val /= test_size;
```

Comment:
Is it reasonable to insert assert for checking `test_size` equal zero?

Reply:
Yes, thank you, I will add assert here